examples/swap_networks.py

```diff
@@ -0,0 +1,158 @@
+"""Demonstrates swap networks.
+
+Swap networks are used to get around limited connectivity in a hardware device.
+In this example, we'll give a concrete application by using the acquaintance
+submodule to implement a single-round QAOA for Maximum Cut on a linearly
+connected device even with an arbitrary problem graph.
+
+Specifically, we'd like to implement the circuit U_mix(β) * U_phase(𝛾), where
+
+    U_mix(β) = X**β ⊗ X**β ⊗ ⋯ ⊗ X**β,
+    U_phase(𝛾) = ZZ_{e_m}**𝛾 ⋯ ZZ_{e_2}**𝛾· ZZ_{e_1}**𝛾, and
+    {e_1, e_2, …, e_m} are the edges of the graph whose maximum cut we would
+        like to find.
+
+The details of QAOA are irrelevant; for our purposes here, it is sufficient to
+say that we want to apply a 2-qubit gate to pairs of qubits, even when the
+implied graph is not a subgraph of the hardware adjacency graph.
+"""
+
+import itertools
+import random
+from typing import Dict, Sequence, Tuple, TypeVar, Union
+
+import cirq
+import cirq.contrib.acquaintance as cca
+
+LogicalIndex = TypeVar('LogicalIndex', int, cirq.Qid)
+LogicalIndexSequence = Union[Sequence[int], Sequence[cirq.Qid]]
+LogicalGates = Dict[Tuple[LogicalIndex, ...], cirq.Gate]
+LogicalMappingKey = TypeVar('LogicalMappingKey', bound=cirq.Qid)
+LogicalMapping = Dict[LogicalMappingKey, LogicalIndex]
+
+
+def get_random_graph(n_vertices, edge_prob=0.5):
+    return [
+        ij for ij in itertools.combinations(range(n_vertices), 2)
+        if random.random() <= edge_prob
+    ]
+
+
+def get_phase_sep_circuit(gates: LogicalGates,
+                          qubits: Sequence[cirq.Qid],
+                          initial_mapping: LogicalMapping,
+                          verbose: bool = True):
+    # An acquaintance strategy containing permutation gates and acquaintance
+    # opportunity gates.
+    circuit = cca.complete_acquaintance_strategy(qubits, 2)
+    if verbose:
+        print('Circuit containing a single complete linear swap network:')
+        print(circuit)
+        print('\n\n')
+
+    acquaintance_opportunities = cca.get_logical_acquaintance_opportunities(
+        circuit, initial_mapping)
+    assert set(frozenset(edge) for edge in gates) <= acquaintance_opportunities
```

Comment:
@babbush I'm hesitant to make GreedyExecutionStrategy assert that it has inserted every logical gate, because there might be use cases in which that's not what you want, but this does essentially the same thing, by asserting that it has the opportunity to insert every logical gate.